File: project2/week1/Hyungjun/ImplementTrie.go
type Trie struct {
	children map[rune]*Trie
	isLast bool
}

func Constructor() Trie {
	return Trie{
		children: make(map[rune]*Trie),
		isLast: false,
	}
}

func (this *Trie) Insert(word string) {
	trie := this
	for _, letter := range word {
		if trie.children[letter] == nil {
			trie.children[letter] = &Trie{
				children: make(map[rune]*Trie),
				isLast: false,
			}
		}
		trie = trie.children[letter]
	}
	trie.isLast = true
}

func (this *Trie) Search(word string) bool {
	trie := this
	for _, letter := range word {
		if _, exist := trie.children[letter]; !exist {
			return false
		}
		trie = trie.children[letter]
	}
	
	return trie.isLast
}

func (this *Trie) StartsWith(prefix string) bool {
	trie := this
	for _, letter := range prefix {
		if _, exist := trie.children[letter]; !exist {
			return false
		}
		trie = trie.children[letter]
	}
	return true
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */

File: project2/week1/Eunchong/implementTrie.cc
class Trie {
public:
    struct Node {
        Node* tn[26];
        bool wordEnd;

        Node(): wordEnd(false) {
            for (int i = 0; i < 26; i++) {
                tn[i] = NULL;
            }
        }
    };

    Trie() {
        root = new Node();
    }
    
    void insert(string word) {
        Node* cur = root;

        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';
            if (cur->tn[idx] == NULL) cur->tn[idx] = new Node();
            cur = cur->tn[idx];
        }
        cur->wordEnd = true;
    }
    
    bool search(string word, bool prefix = false) {
        Node* cur = root;

        for (int i = 0; i < word.size(); i++) {
            int idx = word[i] - 'a';
            if (cur->tn[idx] == NULL) return false;
            cur = cur->tn[idx];
        }
        if (prefix = false) return cur->wordEnd;
        return true;
    }
    
    bool startsWith(string prefix) {
        return search(prefix, true);
    }

private:
    Node* root;
};

/**
 * Your Trie object will be instantiated and called as such:
 * Trie* obj = new Trie();
 * obj->insert(word);
 * bool param_2 = obj->search(word);
 * bool param_3 = obj->startsWith(prefix);
 */
